diag_manager/fms_diag_buffer.F90

```diff
@@ -0,0 +1,1271 @@
+            deallocate(buffobj%buffer)
+        type is (buffer2d)
+            deallocate(buffobj%buffer)
+        type is (buffer3d)
+            deallocate(buffobj%buffer)
+        type is (buffer4d)
+            deallocate(buffobj%buffer)
+        type is (buffer5d)
+            deallocate(buffobj%buffer)
+    end select
+    deallocate(buffobj%missing_value)
+    deallocate(buffobj%buffer_id)
+    deallocate(buffobj%interp_method)
+    deallocate(buffobj%tile_count)
+    deallocate(buffobj%area)
+    deallocate(buffobj%volume)
+end subroutine
+
+!! -----------Type-specific routines for buffer0-5d 
+
+!! allocations could be done in one routine for 0-5d if buffobj is changed to fmsDiagBuffer_class
+!! not sure which approach would be better
+
+!> allocates a scalar buffer to given mold type
+!> @returns buffer id for 
+integer function allocate_buffer_0d(buffobj, mold, init_time) &
+result(rslt)
+    class(buffer0d), intent(inout), target :: buffobj !< scalar buffer object
+    class(*),intent(in) :: mold !< allocates to the type of mold
+    type(time_type), intent(in), optional :: init_time
+    type(fmsDiagBufferContainer_type), allocatable :: buff_type
+
+    select type (mold)
+        type is (integer(kind=i4_kind))
+            allocate(integer(kind=i4_kind) :: buffobj%buffer)
+            allocate(integer(kind=i4_kind) :: buffobj%missing_value)
+            allocate(integer(kind=i4_kind) :: buffobj%data_RANGE(2))
+        type is (integer(kind=i8_kind))
+            allocate(integer(kind=i8_kind) :: buffobj%buffer)
+            allocate(integer(kind=i8_kind) :: buffobj%missing_value)
+            allocate(integer(kind=i8_kind) :: buffobj%data_RANGE(2))
+        type is (real(kind=r4_kind))
+            allocate(real(kind=r4_kind) :: buffobj%buffer)
+            allocate(real(kind=r4_kind) :: buffobj%missing_value)
+            allocate(real(kind=r4_kind) :: buffobj%data_RANGE(2))
+        type is (real(kind=r8_kind))
+            allocate(real(kind=r8_kind) :: buffobj%buffer)
+            allocate(real(kind=r8_kind) :: buffobj%missing_value)
+            allocate(real(kind=r8_kind) :: buffobj%data_RANGE(2))
+        class default
+             call mpp_error("allocate_buffer_0d", &
+                     "The mold value passed to allocate a buffer is not a r8, r4, i8, or i4",&
+                     FATAL)
+    end select
+    allocate(buffobj%buffer_id)
+    allocate(buffobj%interp_method)
+    allocate(buffobj%tile_count)
+    allocate(buffobj%area)
+    allocate(buffobj%volume)
+
+    num_buffers(0) = num_buffers(0) + 1
+    rslt = num_buffers(0)
+    buffobj%buffer_id = num_buffers(0)
+
+end function
+
+!> allocates a 1D buffer to given mold type
+integer function allocate_buffer_1d(buffobj, mold, size) &
+result(rslt)
+    class(buffer1d), intent(inout), target :: buffobj !< scalar buffer object
+    class(*),intent(in) :: mold !< allocates to the type of mold
+    integer, intent(in) :: size !< dimension bounds
```

Comment:
you shouldn't use size as a variable name.  size() is an intrinsic function.  I use `isize` personally.